Libnd4j - "unique" op - incorrect index values

The first output is correct; the second is wrong.
The second output has same length as the input, and is the index of that value into the "unique" array.

This also impacts unique_with_counts in the same way (outputs 0 and 2 seem correct for unique_with_counts)

See also: https://www.tensorflow.org/versions/r1.8/api_docs/python/tf/unique

```
    @Test
    public void testUnique(){
        INDArray in = Nd4j.trueVector(new double[]{3, 4, 3, 1, 3, 0, 2, 4, 2, 4});

        INDArray expUnique = Nd4j.trueVector(new double[]{3, 4, 1, 0, 2});
        INDArray expUniqueIdxs = Nd4j.trueVector(new double[]{0, 1, 0, 2, 0, 3, 4, 1, 4, 1});

        INDArray outUnique = Nd4j.create(expUnique.shape());
        INDArray outUniqueIdxs = Nd4j.create(expUniqueIdxs.shape());

        DynamicCustomOp op = DynamicCustomOp.builder("unique")
                .addInputs(in)
                .addOutputs(outUnique, outUniqueIdxs)
                .build();

        String err = OpValidation.validate(new OpTestCase(op)
                .expectedOutput(0, expUnique)
                .expectedOutput(1, expUniqueIdxs));

        assertNull(err);
    }
```

```
java.lang.AssertionError: expected null, but was:<Output 1 failed: INDArray equality failed:
Expected:
[         0,    1.0000,         0,    2.0000,         0,    3.0000,    4.0000,    1.0000,    4.0000,    1.0000]
Actual:
[         0,    1.0000,         0,    3.0000,         0,    5.0000,    6.0000,    1.0000,    6.0000,    1.0000]>
```

--- a/libnd4j/include/ops/declarable/helpers/cpu/unique.cpp
+++ b/libnd4j/include/ops/declarable/helpers/cpu/unique.cpp
@@ -71,9 +71,10 @@ namespace helpers {
                 (*counts)(e) = countsMap[valuesVector[e]];
         }
 
-#pragma omp parallel for if(indices->lengthOf() > Environment::getInstance()->elementwiseThreshold()) schedule(static)
+//#pragma omp parallel for if(indices->lengthOf() > Environment::getInstance()->elementwiseThreshold()) schedule(static)
         for (int e = 0; e < indices->lengthOf(); e++) {
-            (*indices)(e) = indicesMap[(*input)(e)];
+            auto posI = std::find(valuesVector.begin(), valuesVector.end(), (*input)(e));
+            (*indices)(e) = std::distance(valuesVector.begin(), posI);//indicesMap[(*input)(e)];
         }
 
         return ND4J_STATUS_OK;

--- a/libnd4j/tests_cpu/layers_tests/DeclarableOpsTests10.cpp
+++ b/libnd4j/tests_cpu/layers_tests/DeclarableOpsTests10.cpp
@@ -132,31 +132,19 @@ TEST_F(DeclarableOpsTests10, Test_Size_at_1) {
 
     delete result;
 }
-/*
+
 ////////////////////////////////////////////////////////////////////////////////
-TEST_F(DeclarableOpsTests10, ReverseMod_BP_Test_1) {
+TEST_F(DeclarableOpsTests10, Unique_SGO_Test_1) {
+    NDArray<double> input({3., 4., 3., 1., 3., 0., 2., 4., 2., 4.});
+    NDArray<double> expIdx({0., 1., 0., 2., 0., 3., 4., 1., 4., 1.});
+    NDArray<double> exp({3., 4., 1., 0., 2.});
 
-    NDArray<double> inputX('c', {1, 6});
-    NDArray<double> inputY('c', {3, 4, 5, 1});
-
-    NDArray<double> axis(1.);
-
-    NDArray<double> gradO('c', {3, 4, 5, 6});
-
-    int exclusive, reverse;
-    inputX.assign(2.0);
-    inputY.assign(9.0);
-    ////////////////////////////////////////
-
-
-    const OpArgsHolder<double> argsHolderFF({&inputX, &inputY}, {}, {});
-    const OpArgsHolder<double> argsHolderBP({&inputX, &inputY, &gradO}, {}, {});
-
-    nd4j::ops::reversemod<double> opFF;
-    nd4j::ops::reversemod_bp<double> opBP;
-
-    const bool isGradCorrect = GradCheck::checkGrad(opFF, opBP, argsHolderFF, argsHolderBP);
-
-    ASSERT_TRUE(isGradCorrect);
+    nd4j::ops::unique<double> op;
+    auto res = op.execute({&input}, {}, {});
+    ASSERT_TRUE(res->status() == ND4J_STATUS_OK);
+    //res->at(0)->printIndexedBuffer("Unique values");
+    //res->at(1)->printIndexedBuffer("Unique idxs");
+    ASSERT_TRUE(exp.equalsTo(res->at(0)));
+    ASSERT_TRUE(expIdx.equalsTo(res->at(1)));
+    delete res;
 }
-*/

--- a/libnd4j/tests_cpu/layers_tests/DeclarableOpsTests3.cpp
+++ b/libnd4j/tests_cpu/layers_tests/DeclarableOpsTests3.cpp
@@ -105,7 +105,7 @@ TEST_F(DeclarableOpsTests3, Test_Permute_2) {
 TEST_F(DeclarableOpsTests3, Test_Unique_1) {
     NDArray<float> x('c', {1, 5}, {1, 2, 1, 2, 3});
     NDArray<float> expV('c', {3}, {1, 2, 3});
-    NDArray<float> expI('c', {5}, {0, 1, 0, 1, 4});
+    NDArray<float> expI('c', {5}, {0, 1, 0, 1, 2});
 //    NDArray<float> expI('c', {3}, {0, 1, 4});
 
     nd4j::ops::unique<float> op;
@@ -129,7 +129,7 @@ TEST_F(DeclarableOpsTests3, Test_Unique_1) {
 TEST_F(DeclarableOpsTests3, Test_Unique_2) {
     NDArray<float> x('c', {1, 5}, {1, 2, 1, 2, 3});
     NDArray<float> expV('c', {3}, {1, 2, 3});
-    NDArray<float> expI('c', {5}, {0, 1, 0, 1, 4});
+    NDArray<float> expI('c', {5}, {0, 1, 0, 1, 2});
     NDArray<float> expC('c', {3}, {2, 2, 1});
 
     nd4j::ops::unique_with_counts<float> op;

